Source/WebGPU/WebGPU/BindGroupLayout.mm

```diff
@@ -173,24 +195,47 @@ static bool containsStage(WGPUShaderStageFlags stageBitfield, auto stage)
     for (size_t i = 0; i < stageCount; ++i)
         arguments[i] = [NSMutableArray arrayWithCapacity:descriptor.entryCount];
 
-    Vector<WGPUBindGroupLayoutEntry> bindGroupLayoutEntries;
-    bindGroupLayoutEntries.resize(descriptor.entryCount);
+    Vector<BindGroupLayout::Entry> bindGroupLayoutEntries;
+    bindGroupLayoutEntries.reserveInitialCapacity(descriptor.entryCount);
     for (uint32_t i = 0; i < descriptor.entryCount; ++i) {
         const WGPUBindGroupLayoutEntry& entry = descriptor.entries[i];
-        if (entry.nextInChain)
-            return BindGroupLayout::createInvalid(*this);
+        if (entry.nextInChain) {
+            if (entry.nextInChain->sType != static_cast<WGPUSType>(WGPUSTypeExtended_BindGroupLayoutEntryExternalTexture))
+                return BindGroupLayout::createInvalid(*this);
+            if (entry.nextInChain->next)
+                return BindGroupLayout::createInvalid(*this);
+        }
 
         shaderStageForBinding.add(entry.binding + 1, entry.visibility);
         MTLArgumentDescriptor *descriptor = nil;
 
-        if (BindGroupLayout::isPresent(entry.buffer))
-            descriptor = createArgumentDescriptor(entry.buffer);
-        else if (BindGroupLayout::isPresent(entry.sampler))
-            descriptor = createArgumentDescriptor(entry.sampler);
-        else if (BindGroupLayout::isPresent(entry.texture))
-            descriptor = createArgumentDescriptor(entry.texture);
-        else if (BindGroupLayout::isPresent(entry.storageTexture))
-            descriptor = createArgumentDescriptor(entry.storageTexture);
+        BindGroupLayout::Entry::BindingLayout bindingLayout;
+        auto processBindingLayout = [&](const auto& type) {
+            if (!BindGroupLayout::isPresent(type))
+                return true;
+            if (descriptor)
+                return false;
+            descriptor = createArgumentDescriptor(type);
+            if (!descriptor)
```

Comment:
`!descriptor` is already checked on line 195, so it is duplicated with this additional code line here.

I don't think it is sufficient to remove it from line 195 however, since none of the `if` conditionals may be true unless the caller guarantees that

Reply:
I think both are necessary.

As you say, the one on line 195 is necessary in case none of the `if`s are taken.

The one here is necessary to make sure that, if we try to create an argument descriptor of one particular type, and we fail, then we shouldn't even try to create an argument descriptor of another type. A single failure should tank the whole thing, rather than saying "well one entry is invalid but another one was valid so let's use the valid one"